Issue #377: Video remove button

aforcier opened the pull request:
Fixes #377 and #408.

Adds a permanent span container to videos, with a delete button:

![377-video-delete-overlay](https://cloud.githubusercontent.com/assets/9613966/15987672/00d6a184-2fff-11e6-9495-cea95f943901.png)

This PR also fixes an issue reported by @tonyr59h, that adding remote videos to an empty post makes it difficult to tap under the video afterward. The problem was that the editor wasn't adding a newline after those videos, like it does for all images and for uploading videos.

The best way to test is to `subtree pull` this branch into WPAndroid, but it's also possible to check the remote video insertion code in isolation by calling [`ZSSEditor.insertVideo()`](https://github.com/wordpress-mobile/WordPress-Editor-Android/blob/531f3687115620564764e0c1a94f5875438cb151/libs/editor-common/assets/ZSSRichTextEditor.js#L1538) in the Chrome dev console with appropriate arguments.

There's a more general issue with paragraph formatting around media, which applies to all media and is due to the newline under media being stripped when switching to HTML mode and back, or opening an existing post. That's being tracked in #409.

Hopefully this also fixes all issues with #296 (needs testing on the Z5 and the HTC One now that `span`-wrapped videos are in the picture).

cc @maxme, and also @tonyr59h for a runthrough since he's the most talented at summoning video bugs

aforcier committed:
Show delete button overlay for videos that have finished uploading

aforcier committed:
Add delete button overlay for videos added directly (without being uploaded)

- Due to paragraph/cursor issues with video nodes, we need to first insert the video, and then wrap it in the container span

aforcier committed:
Fixed paragraph issues when adding remote videos at the start of a post

aforcier committed:
Track video edit containers for mutation

- This fixes an issue where manually deleting videos (by backspacing) can cause the cursor to lose focus

aforcier committed:
Added a null check to ZSSEditor.sendEnabledStyles

aforcier committed:
Fixed trackNodeForMutation usage when video upload completes

aforcier committed:
Remove edit containers wrapping videos when stripping visual formatting

aforcier committed:
Add edit containers to videos when applying visual formatting

aforcier committed:
When adding edit containers for videos in ZSSEditor.applyVisualFormatting, track them for mutation

maxme commented:
The button looks good and works well but apparently it breaks the playback (pulled the subtree in wpandroid): https://cloudup.com/c6RwYZBACc8

maxme commented:
It works better on the example app https://cloudup.com/cAf7fHN8VyA but in both tests, I had the cursor moved before the X button after tapping the video.

tonyr59h commented:
> I had the cursor moved before the X button after tapping the video.

Saw the same in the example app.

aforcier committed:
Extend contenteditable=false to the entire edit container for videos

- Fixes an issue where the cursor would sometimes appear next to the close button when playing a video

aforcier commented:
> The button looks good and works well but apparently it breaks the playback (pulled the subtree in wpandroid): https://cloudup.com/c6RwYZBACc8

This turns out to be a pre-existing issue: https://github.com/wordpress-mobile/WordPress-Editor-Android/issues/408.

> It works better on the example app https://cloudup.com/cAf7fHN8VyA but in both tests, I had the cursor moved before the X button after tapping the video.

This should be fixed in 0acaa82 - I extended `contenteditable="false"` to apply to the whole container, not just the delete overlay. With that change I can't reproduce the issue anymore.

Ready for another look @maxme 😄

aforcier committed:
Fixed #408

- Stops continuous reload attempts for unprocessed videopress videos when opening a post or switching to HTML mode and back

aforcier commented:
Updated with a fix for  #408 (flickering videopress videos when switching to HTML mode and back right after uploading a videopress video).

maxme commented:
Looks good, one issue noticed https://github.com/wordpress-mobile/WordPress-Editor-Android/issues/415 but it's not related to this changeset.

:shipit:
